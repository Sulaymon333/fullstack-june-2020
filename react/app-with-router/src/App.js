import React from 'react';
import { Link, Redirect, Route, Switch } from 'react-router-dom'
import Category from './Category'
import Products from './Products'
import Login, {fakeAuth} from './Login'


const Home = () => (
  <div>
    <h2>Home</h2>
  </div>
  )

const Admin = ( {match}) => {
  return (
    <div>
    <h2>Welcome Admin</h2>
  </div>
  )
}

const PrivateRoute = ({ component: Component, ...rest }) => {
  return (
    <Route
      {...rest}
      render={props =>
        fakeAuth.isAuthenticated === true ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{ pathname: "/login", state: { from: props.location } }}
          />
        )
      }
    />
  );
};

function App() {
  return (
    <div>
      <nav>
        <ul>
          <li>
            <Link to="/">Home</Link>
          </li>
          <li>
            <Link to="/category">Category</Link>
          </li>
          <li>
            <Link to="/products">Products</Link>
          </li>
          <li>
            <Link to="/admin">Admin Area</Link>
          </li>
        </ul>
      </nav>

      {/* Route components rendered if following path prop matches URL */}
      <Switch>
        <Route path="/login" component={Login} />
        <Route exact path="/" component={Home} />
        <Route path="/category" component={Category} />
        <PrivateRoute path="/admin" component={Admin} />
        <Route path="/products" component={Products} />
      </Switch>
    </div>
  );
}

export default App;
