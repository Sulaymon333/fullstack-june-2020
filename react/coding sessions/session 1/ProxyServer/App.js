const express = require('express')
const axios  = require('axios')
var cors = require('cors')
require('dotenv').config()
const app = express()
const TOKEN = process.env.TOKEN

app.use(cors())

app.get('/', (req,res)=>{
    console.log('hello World')
    res.send('<h1>Hello oooh</h1>')
})

app.get('/restaurants', async (req, res)=>{
    console.log('/api in')
    

    if(!req.query || !req.query.searchTerm || !req.query.location){
        return res.status(400).end('searchTerm and location are the required query string parameters!')
    }
        const searchTerm = req.query.searchTerm
        const location = req.query.location

        console.log('searchTerm is ', searchTerm)
        console.log('location is ', location)

        let config ={
            method:'get',
            url:`https://api.yelp.com/v3/businesses/search?term=${searchTerm}&location=${location}`,
            headers: {
              'Authorization': `${TOKEN}`
            }
        }
        const response = await axios(config)
        return res.send(response.data)
    
})



const PORT=3001
app.listen(PORT, ()=> {
    console.log(`Server running on port ${PORT}`)
})