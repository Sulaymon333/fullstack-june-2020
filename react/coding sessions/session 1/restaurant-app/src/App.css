.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}
input{
  margin: 20px;
  padding: 10px;
  width: 300px;
  border: 2px solid black;
}
.restaurants {
  display: flex;
  margin: 20px;
  flex-direction: row;
  flex-wrap: wrap;
}
.cards {
  border: 2px solid black;
  padding: 10px;
  margin: 10px;
  width:300px;
  height: 300px;
  text-align: center;
  align-items: flex-start;
 
}
h1 {
  padding: 10px;
  margin: 10px;
  text-align: center;
}
.restaurant{
  font-weight: bold;
}
.restaurants img {
  width:200px;
  height: 200px;
}
@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.nav-bar{
  min-width: 100%;
  background-color: #0d86cb;
  display: flex;
  justify-content: space-around;
  align-items: center;
  color:white;
  position: sticky;
  top: 0;
}
.menu{
  background-color: white;
  min-width: 15%;
  min-height: 100vh;
  z-index: 10;
  color: #0d86cb;
  position: fixed;
  top: 0;
  left: 0;
}

.fa-heart{
  float: left;
  cursor: pointer;
}