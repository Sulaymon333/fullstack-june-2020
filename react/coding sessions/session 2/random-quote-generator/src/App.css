.container{
  background-color: black;
  color: white;
  text-align: center;
  margin: auto;
  width:100%;
  height: 100vh;
  padding: 30px;;
}

.wrapper{
  padding: 10px;
 margin-top:200px;
 margin-left: 200px;
 margin-right: 200px;
  border: 2px solid whitesmoke;
}

.quote{
  font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;
  font-size: large;
}

.author{
  font-style: italic;
}

.button-main{
  padding: 10px;
  margin: 30px;
}

.button-main:hover{
  opacity: 0.7;
}
.fade-in {
  animation: fadeIn ease 20s;
  -webkit-animation: fadeIn ease 20s;
  -moz-animation: fadeIn ease 20s;
  -o-animation: fadeIn ease 20s;
  -ms-animation: fadeIn ease 20s;
  }