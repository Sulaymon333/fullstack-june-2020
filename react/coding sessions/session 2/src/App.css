.container {
  text-align: center;
  background-color: black;
  color: white;
  width: 100%;
  height: auto;
  margin:auto;
  padding: 100px;
  
}

.quote {
  font-family: cursive;
  font-size: large;
  margin: 20px;
}

.author {
  text-align: right;
  font-style: italic;
  margin: 20px;
}

.wrapper {

  padding: 20px;
  align-self: center;
  margin:auto;
  text-align: center;
  width: 500px;
  height: 500px;
  border: 4px solid white
}

button{
  padding: 10px;
  margin:10px;
}