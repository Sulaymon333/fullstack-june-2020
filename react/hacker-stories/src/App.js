import React from 'react';
import axios from 'axios'

import styled from 'styled-components'

import searchTerm from './SearchForm'
import List from "./List";

import { ReactComponent as Check} from './check.svg'


// import styles from './App.module.css'
// import cs from 'classnames'

// const initialStories = [
//   {
//     title: 'React',
//     url: 'https://reactjs.org/',
//     author: 'Jordan Walke',
//     num_comments: 3,
//     points: 4,
//     objectID: 0,
//   },
//   {
//     title: 'Redux',
//     url: 'https://redux.js.org/',
//     author: 'Dan Abramov, Andrew Clark',
//     num_comments: 2,
//     points: 5,
//     objectID: 1,
//   },
// ];

// A to fetch popular tech stories for a certain query
const API_ENDPOINT = 'http://hn.algolia.com/api/v1/search?query='


// const getAsyncStories = () =>
//   new Promise((resolve, reject) => setTimeout(reject, 2000) );

// Styled components
const StyledContainer = styled.div` 
height: 100vw;
padding: 20px;

background: #83a4d4;
background: linear-gradient(to left, #b6fbff, #83a4d4);

color: #171212`

const StyledHeadlinePrimary = styled.h1`
font-size: 45px;
font-weight: 300;
letter-spacing: 2px;
`

const StyledItem = styled.div`
  display: flex;
  align-items: center;
  padding-bottom: 5px;
`
const StyledColumn = styled.span`
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  a {
  color: inherit;
  }

  width: ${props => props.width}
`

const StyledButton = styled.button`
  background: transparent;
  border: 1px solid #171212;
  padding: 5px;
  transition: all 0.1s ease-in;

  &:hover >svg > g {
  fill: #ffffff;
  stroke: #ffffff;
  }
`

const StyledButtonSmall = styled(StyledButton)`
  padding: 5px;
`

const StyledButtonLarge = styled(StyledButton)`
  padding: 10px;
`

const StyledSearchForm = styled.form`
  padding: 10px 0 20px 0;
  display: flex;
  align-items: baseline;
`

const StyledLabel = styled.label`
  border-top: 1px solid #171212;
  border-bottom-style: 1px solid #171212;
  padding-left: 5px;
  font-size: 24px;
`

const StyledInput = styled.input`
  border: none;
  border-bottom: 1px solid #171212;
  background-color: transparent;
  font-size: 24px;
`

// Function for app 

const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );

  React.useEffect(() => {
    localStorage.setItem(key, value);
  }, [value, key]);

  return [value, setValue];
};

const storiesReducer = (state, action) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
      isLoading: true,
      isError: false,
      }
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
      isLoading: false,
      isError: false,
      data: action.payload,
      }
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
      isLoading: false,
      isError: true,
      }
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
          story => action.payload.objectID !== story.objectID
        )
      }
    default:
      throw new Error();
  }
};

const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState(
    'search',
    'React'
  );

  const [url, setUrl] = React.useState(
    `${API_ENDPOINT}${searchTerm}`
  )

  const [stories, dispatchStories] = React.useReducer(
    storiesReducer,
    { data: [], isLoading: false , isError: false }
  );
  // const [isLoading, setIsLoading] = React.useState(false);
  // const [isError, setIsError] = React.useState(false);

  // const handleFetchStories = React.useCallback(() => {
  //   dispatchStories({type: 'STORIES_FETCH_INIT'});

  //   fetch(url) // native browser's fetch API to make request
  //     .then(response => response.json()) // Translating into JSON

  //     .then(result => {
  //       dispatchStories({
  //         type: 'STORIES_FETCH_SUCCESS',
  //         payload: result.hits
  //       })
  //     })

  //     .catch(() => 
  //       dispatchStories({type: 'STORIES_FETCH_FAILURE'})
  //     )
  //   }, [url])

  const handleFetchStories = React.useCallback(async() => {
    dispatchStories({type: 'STORIES_FETCH_INIT'})

    try {
      const result = await axios.get(url)
      .then(result => {
        dispatchStories({
          type: 'STORIES_FETCH_SUCCESS',
          payload: result.data.hits
        })
      })

    } catch {
      dispatchStories({type: 'STORIES_FETCH_FAILURE'})
    }
  }, [url])
 
  React.useEffect(() => {
    handleFetchStories()
  }, [handleFetchStories])

  const handleRemoveStory = item => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  };

  const handleSearch = event => {
    setSearchTerm(event.target.value);
  };

  const handleSearchInput = event => {
    setSearchTerm(event.target.value)
  }

  const handleSearchSubmit = event => {
    setUrl(`${API_ENDPOINT}${searchTerm}`)
    event.preventDefault()
  }

  return (
    <StyledContainer>
      <StyledHeadlinePrimary>My Hacker Stories</StyledHeadlinePrimary>

      <SearchForm
        searchTerm = {searchTerm}  
        onSearchInput = {handleSearchInput}
        onSearchSubmit = {handleSearchSubmit}
      />

      {stories.isError && <p>Something went wrong ...</p>}

      {stories.isLoading ? (
        <p>Loading ...</p>
      ) : (
        <List
          list={stories.data}
          onRemoveItem={handleRemoveStory}
        />
      )}
    </StyledContainer>
  );
};

export default App;
export { storiesReducer, SearchForm, InputWithLabel, List, Item }
