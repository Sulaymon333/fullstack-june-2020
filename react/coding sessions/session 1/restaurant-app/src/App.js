/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useEffect } from "react";
import axios from "axios";

import "./App.css";

import DisplayRestaurant from "./components/DisplayRestaurant";
import Navbar from "./components/Navbar";

const App = () => {
  const [restaurantList, setRestaurantList] = useState({ businesses: [] });
  const [search, setSearch] = useState("restaurants");
  const [favorite, setFavorite] = useState(false)
  const [ isLoading, setisLoading] = useState(false)

  const fetchingData = async () => {
    setisLoading(true)
    const url = `http://localhost:3001/restaurants?searchTerm=${search}&location=helsinki`;
    const response = await axios.get(url);
    // console.log("url is ", url);
    // console.log("response is ", response.data);
    response.data.businesses.forEach( res => Object.assign(res,{favorite: false}))
    setRestaurantList(response.data);
    setisLoading(false)
  };

  useEffect(() => {
    // get data from localStorage
    let restaurantData = localStorage.getItem('restaurantList')
    // convert it from JSON file
    restaurantData = JSON.parse(restaurantData)
    console.log('data from localstorage',restaurantData)

    if(restaurantData !== null){
      setRestaurantList({ businesses: [...restaurantData]})
    }else{
      fetchingData();
    }
  }, [search]);


  // filter out the items its favorite is true
  const favoriteList = restaurantList.businesses.filter( restaurant => restaurant.favorite === true)

  const handleClick = () => {
    let sortedRestaurants = restaurantList.businesses.sort((a, b) => {
      return b.rating - a.rating;
    });
    setRestaurantList({businesses:[...sortedRestaurants]});
    console.log("result is ", sortedRestaurants);
  };


  const searchRestaurants = (e) => {
    e.preventDefault();
    console.log("button clicked");

    const newSearch = document.getElementById("addSearch");
    if (newSearch.value !== "") {
      console.log("data from newSearch is ", newSearch.value);
      setSearch(newSearch.value);
    }
  };

  // console.log(restaurantList)
  // console.log(favorite)

  return (
    <div className="Main-container">
      <Navbar searchRestaurants={searchRestaurants} setFavorite={setFavorite} />

      {/* try to use the spinner when data takes time to load */}
      {isLoading && <div className="loader">Loading...</div>}

      {restaurantList && 
      <DisplayRestaurant 
        // send favoriteList as a props to this component when the favorite state is true
        restaurantList={favorite ? favoriteList : restaurantList.businesses} 
      />}
    </div>
  );
};

export default App;