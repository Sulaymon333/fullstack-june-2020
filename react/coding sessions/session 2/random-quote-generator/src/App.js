import React, { useState } from 'react';
import Quotes from './QuotesDatabase';
import './App.css';

const App = () => {
  const [quote, setQuote] = useState(Quotes[0].quote)
  const [author, setAuthor] = useState(Quotes[0].author)

  const handleClick = () => {
    console.log('Button Click')
    const rand =Math.floor(Math.random()* Quotes.length)
    console.log('Random Number is ', rand)
    setQuote(Quotes[rand].quote)
    setAuthor(Quotes[rand].author)
  }
 
  
  return(
    <div className="container">
        <div className="wrappers w3-animate-top">
          <p className="quote">{quote}</p>
          {author ? <p className="author">- {author}</p> : <p className="author">- Unknown</p>}
        </div>
        <button className="button-main" onClick={handleClick}>Generate Quote</button>
    </div>
  )
}

export default App;
