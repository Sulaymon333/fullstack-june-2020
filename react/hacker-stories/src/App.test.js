import React from 'react';
import { 
  render,
  screen,
  fireEvent,
  act
} from '@testing-library/react';
import App, { storiesReducer, SearchForm, InputWithLabel, List, Item } from './App';

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  }

const storyTwo =  {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
  }

const stories = [storyOne, storyTwo]

//// Test suites and test cases examples
// describe('App Component', () => {
//   it('removes an item when clicking the dismiss', () =>{

//   })

//   it('requests some initial stories from an API', () => {

//   })
// })

// describe('something truthy and falsy', () => {
//   test('true to be true', () => {
//     expect(true).toBeTruthy()
//   })

//   test('false to be false', () => {
//     expect(false).toBeFalsy()
//   })
// })

// // Function unit test with Jest
// describe('storiesReducer', () => {
//   test('removes a story from all stories', () => {
//     const action = { type: 'REMOVE_STORY', payload: storyOne}
//     const state = {data: stories, isLoading: false, isError: false}

//     const newState = storiesReducer(state, action)

//     const expectedState = {data: [storyTwo], 
//       isLoading: false, 
//       isError: false
//     }

//     expect(newState).toStrictEqual(expectedState)
//   })
// })

// // Item component unit test
// describe('Item', () => {
//   test('renders all properties' ,() => {
//     render(<Item item={storyOne} />)

//     expect(screen.getByText('Jordan Walke')).toBeInTheDocument()
//     expect(screen.getByText('React')).toHaveAttribute(
//       'href',
//       'https://reactjs.org/' 
//     )
//   })

//   // test('renders a clickable dismiss button', () => {
//   //   render(<Item item={storyOne} />)

//   //   expect(screen.getByRole('span').toBeInTheDocument())
//   // })

//   test('clicking the dismiss button calls callback handler', () => {
//     const handleRemoveItem = jest.fn()

//     render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />)

//     fireEvent.click(screen.getByRole('button'))

//     expect(handleRemoveItem).toHaveBeenCalledTimes(1)
//   })
// })

// SearchFrom component using InputWithLabel as child component
describe('SearchForm', () => {
  const searchFormProps = {
    searchTerm: 'React',
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn()
  }

  test('renders the input field with its value', () => {
    render(<SearchForm {...searchFormProps} />)

    expect(screen.getByDisplayValue('React')).toBeInTheDocument()
  })

  test('renders the correct label', () => {
    render(<SearchForm {...searchFormProps } />)

    expect(screen.getByLabelText(/Search/)).toBeInTheDocument()
  })

  test('calls onSearchInput on the input field change', () => {
    render(<SearchForm {...searchFormProps} />)

    fireEvent.change(screen.getByDisplayValue('React'), {
      target: {value: 'Redux'}
    })

    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1)
  })

  test('calls onSearchSubmit on button submit click', () => {
    render(<SearchForm {...searchFormProps} />)

    fireEvent.submit(screen.getByRole('button'))

    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1)
  })

  test('renders snapshot', () => {
    const { container } = render(<SearchForm {...searchFormProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})