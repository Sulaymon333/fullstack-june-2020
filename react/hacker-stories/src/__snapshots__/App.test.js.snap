// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders snapshot 1`] = `
<form
  class="sc-Axmtr cERexT"
>
  <label
    class="sc-AxmLO kXGgS"
    for="search"
  >
    Search:
  </label>
  Â 
  <input
    class="sc-fzozJi dxjiOm"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="sc-AxhUy sc-AxheI fYYzPy"
    type="submit"
  >
    Submit
  </button>
</form>
`;
